src/Simplex/Chat/Library/Commands.hs

```diff
@@ -2928,53 +2946,47 @@ processChatCommand' vr = \case
                 ( CRInvitationUri crData {crScheme = SSSimplex} e2e,
                   CRInvitationUri crData {crScheme = simplexChat} e2e
                 )
-    connectViaContact :: User -> IncognitoEnabled -> CreatedLinkContact -> Maybe SharedMsgId -> Maybe (SharedMsgId, MsgContent) -> Maybe AttachConnToContactOrGroup -> CM ChatResponse
-    connectViaContact user@User {userId} incognito (CCLink cReq@(CRContactUri crData@ConnReqUriData {crClientData}) sLnk) welcomeSharedMsgId msg_ attachConnTo_ = withInvitationLock "connectViaContact" (strEncode cReq) $ do
+    connectViaContact :: User -> Maybe PreparedChatEntity -> IncognitoEnabled -> CreatedLinkContact -> Maybe SharedMsgId -> Maybe (SharedMsgId, MsgContent) -> CM ConnectViaContactResult
+    connectViaContact user@User {userId} preparedEntity_ incognito (CCLink cReq@(CRContactUri crData@ConnReqUriData {crClientData}) sLnk) welcomeSharedMsgId msg_ = withInvitationLock "connectViaContact" (strEncode cReq) $ do
       let groupLinkId = crClientData >>= decodeJSON >>= \(CRDataGroup gli) -> Just gli
           cReqHash = ConnReqUriHash . C.sha256Hash . strEncode
           cReqHash1 = cReqHash $ CRContactUri crData {crScheme = SSSimplex}
           cReqHash2 = cReqHash $ CRContactUri crData {crScheme = simplexChat}
-      case groupLinkId of
-        -- contact address
+      -- groupLinkId is Nothing for business chats
+      when (isJust msg_ && isJust groupLinkId) $ throwChatError CEConnReqMessageProhibited
+      case preparedEntity_ of
+        Just (PCEContact ct@Contact {activeConn}) -> case activeConn of
+          Nothing -> connect' Nothing cReqHash1 Nothing
+          Just conn@Connection {connStatus, xContactId} -> case connStatus of
+            ConnPrepared -> joinPreparedConn' xContactId conn False
+            _ -> pure $ CVRConnectedContact ct
+        Just (PCEGroup _gInfo GroupMember {activeConn}) -> case activeConn of
+          Nothing -> connect' groupLinkId cReqHash1 Nothing
+          Just conn@Connection {connStatus, xContactId} -> case connStatus of
+            ConnPrepared -> joinPreparedConn' xContactId conn $ isJust groupLinkId
+            _ -> connect' groupLinkId cReqHash1 xContactId -- why not "already connected" for host member?
```

Comment:
as the comment says, it feels like it should be a similar event to contact already exists.